### PR TITLE
[Incremental Imports] Keep top-level type-body hashes away from swiftmodule hashes

### PR DESCRIPTION
**Pull request by davidungar**
(At present includes https://github.com/apple/swift/pull/36133, but I'll rebase after landing it.)

Add `InterfaceHashWithoutTypeMembers` to `BasicSourceFileInfo` and use it when computing swiftmodule hashes.

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
@swift-ci please smoke test

File: lib/Serialization/ModuleFile.cpp [hunk 1]
```diff
@@ -990,9 +990,11 @@ void ModuleFile::collectBasicSourceFileInfo(
 
     BasicSourceFileInfo info;
     info.FilePath = filePath;
-    if (auto fingerprint = Fingerprint::fromString(fpStr))
-      info.InterfaceHashIncludingTypeMembers = fingerprint.getValue();
-    else {
+    if (auto fingerprint = Fingerprint::fromString(fpStr)) {
+      // Only the hash without type members is stored in a module file
+      // because that's what's needed for incremental imports.
+      info.InterfaceHashWithoutTypeMembers = fingerprint.getValue();
```
Comment (rintaro) [inline lib/Serialization/ModuleFile.cpp:996]: This is wrong. `ModuleFileSharedCore.SourceFileListData` is a list of `BasicSourceFileInfo` that is serialized in https://github.com/apple/swift/blob/88d3facc9fa62c40af0df1f2ccb77971529f0992/lib/Serialization/SerializeDoc.cpp#L805-L821
This `fingerprint` is `InterfaceHashIncludingTypeMembers`.

Comment (davidungar) [reply lib/Serialization/ModuleFile.cpp:996]: Thank you very much for pointing this out! I will fix it.

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
Waiting for https://github.com/apple/swift/pull/36168 , then will rebase and fix conflicts with that one.

**Comment by davidungar** [comment]
Closing in favor of https://github.com/apple/swift/pull/36188

**Pull request closed without merging**